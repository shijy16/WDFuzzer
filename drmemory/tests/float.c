/* **********************************************************
 * Copyright (c) 2013-2014 Google, Inc.  All rights reserved.
 * Copyright (c) 2009 VMware, Inc.  All rights reserved.
 * **********************************************************/

/* Dr. Memory: the memory debugger
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation;
 * version 2.1 of the License, and no later version.

 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Library General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

int
main()
{
    char *p1;
    float f;
    double d;
    double da[2];
    int i;

    printf("testing floats on heap\n");
    p1 = (char *) malloc(64);
    f = *((float *)p1); /* error: uninitialized, but won't report until used */
    if (f < 0.f) /* use -> reported */
        d = 4.;
    /* For doubles Dr. Memory will report right away since no shadowing
     * of floating-point registers: except w/ the new i#471 heuristic we need
     * to add a use.
     */
    d = *((double *)(p1+8)); /* error: uninitialized */
    if (d < 0.) /* use -> reported */
        d = 4.;
    *((float*)(p1+16)) = 0.f;
    d = *((double *)(p1+16)); /* error: half uninitialized */
    if (d < 0.) /* use -> reported */
        d = 4.;
    *((float*)(p1+28)) = 0.f;
    d = *((double *)(p1+24)); /* error: other half uninitialized */
    if (d < 0.) /* use -> reported */
        d = 4.;
    /* test unaligned */
    d = *((double *)(p1+34)); /* error: uninitialized */
    if (d < 0.) /* use -> reported */
        d = 4.;
    free((void *)p1);

    printf("testing floats on stack\n");
    da[0] = 0.;
    i = 1; /* if we use da[1] cl identifies the bad read */
    if (da[i] < 0.f) /* error: uninitialized */
        d = 4.;

    printf("all done\n");
    return 0;
}
