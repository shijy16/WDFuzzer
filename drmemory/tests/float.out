# **********************************************************
# Copyright (c) 2011 Google, Inc.  All rights reserved.
# Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
testing floats on heap
testing floats on stack
all done
~~Dr.M~~ ERRORS FOUND:
~~Dr.M~~       0 unique,     0 total unaddressable access(es)
# PR 473614: cl's /RTC1 fills all locals w/ 0xcccccccc
# but /RTC1 is now off (for i#545) so re-enabling
~~Dr.M~~       6 unique,     6 total uninitialized access(es)
~~Dr.M~~       0 unique,     0 total invalid heap argument(s)
~~Dr.M~~       0 unique,     0 total warning(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of leak(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
